File: core/src/main/java/forpleuvoir/dhwu1a/core/message/messagesender/GroupMessageSender.java
package forpleuvoir.dhwu1a.core.message.messagesender;

import com.google.gson.annotations.SerializedName;
import forpleuvoir.dhwu1a.core.common.data.GroupData;
import forpleuvoir.dhwu1a.core.message.base.MessageSender;
import forpleuvoir.dhwu1a.core.user.Group;
import forpleuvoir.dhwu1a.core.user.Member;
import forpleuvoir.dhwu1a.core.user.base.Permission;

import javax.annotation.Nullable;

import static forpleuvoir.dhwu1a.core.common.ApiKey.*;

/**
 * @author forpleuvoir
 * <p>#project_name dhwu1a
 * <p>#package forpleuvoir.dhwu1a.core.message.messagesender
 * <p>#class_name GroupMessageSender
 * <p>#create_time 2021/7/1 21:56
 */
public class GroupMessageSender extends MessageSender {

    @SerializedName(MEMBER_NAME)
    public String memberName;
    @SerializedName(SPECIAL_TITLE)
    public String specialTitle;
    @SerializedName(PERMISSION)
    public Permission permission;
    @SerializedName(JOIN_TIMESTAMP)
    public Integer joinTimestamp;
    @SerializedName(LAST_SPEAK_TIMESTAMP)
    public Integer lastSpeakTimestamp;
    @SerializedName(MUTE_TIME_REMAINING)
    public Integer muteTimeRemaining;
    @SerializedName(GROUP)
    private GroupData group;

    protected GroupMessageSender(Long id) {
        super(id);
    }

    @Nullable
    @Override
    public Group getUser() {
        return group.getUser();
    }

    @Nullable
    public Member getMember() {
        if (group.getUser() != null)
            return group.getUser().getMember(this.id);
        return null;
    }
}

File: core/src/main/kotlin/forpleuvoir/dhwu1a/core/message/messagesender/GroupMessageSender.kt
package forpleuvoir.dhwu1a.core.message.messagesender

import com.google.gson.annotations.SerializedName
import forpleuvoir.dhwu1a.core.common.*
import forpleuvoir.dhwu1a.core.common.data.GroupData
import forpleuvoir.dhwu1a.core.message.base.MessageSender
import forpleuvoir.dhwu1a.core.user.Group
import forpleuvoir.dhwu1a.core.user.Member
import forpleuvoir.dhwu1a.core.user.base.Permission

/**
 * @author forpleuvoir
 *
 * #project_name dhwu1a
 *
 * #package forpleuvoir.dhwu1a.core.message.messagesender
 *
 * #class_name GroupMessageSender
 *
 * #create_time 2021/7/1 21:56
 */
open class GroupMessageSender protected constructor(id: Long) : MessageSender(id) {
    @SerializedName(MEMBER_NAME)
    lateinit var memberName: String
        private set

    @SerializedName(SPECIAL_TITLE)
    lateinit var specialTitle: String
        private set

    @SerializedName(PERMISSION)
    lateinit var permission: Permission
        private set

    @SerializedName(JOIN_TIMESTAMP)
    var joinTimestamp: Int = 0
        private set

    @SerializedName(LAST_SPEAK_TIMESTAMP)
    var lastSpeakTimestamp: Int = 0
        private set

    @SerializedName(MUTE_TIME_REMAINING)
    var muteTimeRemaining: Int = 0
        private set

    @SerializedName(GROUP)
    private lateinit var group: GroupData
    override val user: Group
        get() = group.user
    val member: Member?
        get() = group.user.getMember(this.id)
}